docs/architecture/ErrorReporting.md

```diff
@@ -0,0 +1,144 @@
+# Handling and reporting errors
+
+    WARNING: This document describes future plans for error handling.
+    The UI library does not (yet) implement error handling as described here.
+
+This document describes how errors are handled within the ACS UI library and how they are reported to the surrounding application.
+
+## Error sources
+
+There are two categories of errors, and they are handled differently.
+
+1. ACS backend errors: Errors returned by the REST calls to ACS backend services.
+
+   1. These can be transient or permanent (special case: permission error from REST API).
+
+2. UI errors: Programming errors in the UI library / errors returned from other browser APIs. This is a catch-all for all errors not generated from the ACS SDK libraries.
+
+## ACS backend errors
+
+Handling ACS backend errors is challenging when using either the composites or the pure components with the react bindings provided by the ACS UI library:
+
+* ACS backend API calls are asynchronous
+* ACS backend API calls are often initiated outside of the React component lifecycle methods (e.g on interaction handler callbacks or `useEffect` blocks).
+
+But in both cases, all ACS backend API calls are mediated by the (chat and calling) stateful clients. Thus, the stateful client tees errors to a special field in the state:
+
+```ts
+interface CallClientState {
+    // ... other fields ...
+    errors: { [key: SdkClientMethod]: ErrorState[];
+}
+
+type SdkClientMethod = 'CallClient.createCallAgent' | 'CallClient.getDeviceManager'; // ... and many more
+
+interface ErrorState {
+    /**
+     * Locally generated unique ID for error.
+     * Allows for de-duplicating similar errors from repeated API calls.
+     */
+    id: string;
+    error: Error;
+}
+```
+
+Stateful client API method calls continue to return errors (e.g. as a failed promise). So they can be handled directly as well.
+
+The stateful clients stores the most recent errors per method, evicting older errors if necessary. The number of errors stored can be configured, or set to -1 to disable deletion of older errors. e.g., the composites disable error eviction in the stateful client and delete older errors themselves. By default, only the most recent error is stored.
+
+The errors in the state enables UI library components that act on some errors (e.g., disabling `SendBox` in case no messages can be sent), components that surface UI errors, and adapters that generate error events for the surrounding application.
+
+### Clearing errors
+
+Errors in the state are cleared in one of two ways:
+
+* Application clears the errors via a new Stateful client method.
+
+  ```ts
+  interface StatefulChatClient {
+      // ... Other methods
+      clearErrors(method: SdkClientMethod): void;
+  }
+  ```
+
+* If a previously failed method is retried and succeeds, the stateful client clears the errors on that (and sometimes related) method.
```

Comment:
I would recommend coming up with a POC in our existing library for this. 
Preferably using typescript decorators to make it less intimidating to maintain.

I have a gut feeling that this interception of each method for clearing error state based on the success of its response might end up becoming too complex.